### PR TITLE
016 Templates

### PR DESCRIPTION
**Pull request by qubd**
Here are the template assignments for 201-016. There's a triangle trigonometry assignment missing. Could be added later.

### CODE REVIEW

**Comment by maharper** [comment]
Fantastic Brendan!

We copied the assignments/problems from your 201-016 last term to form the basis of the 201-016 for this term.  This will make it official :-).

There is one gotcha to do with the new template assignment location on the server.  Essentially, we have moved the set defs out of the course `templates` directory.  The reason for the change was because as copies of the assignments and local versions of problems proliferated, it became too difficult to maintain all of the various versions of homework sets and problems.  Following this change there is only a single version of a set def on the server and single versions of linked problems.

The set defs and problem code that lived under `templates` now lives beside the Champlain Problem Library in `Course-Templates`.

You have set your assignments up correctly, so they will live under `model201-016`.  The only catch arises after we make the set defs available in a course.  To do this we put a link from the course `templates` directory pointing to the `model201-<nnn>` directory.  For example, every new NYA course has a link in its `templates` directory called `Template_201-NYA`  This link points to `model201-NYA`.  The set defs under `model201-NYA` show up nicely for import in the Homework Sets Editor and in the Library Browser.  Any OPL or CPL problems in the homework set are found properly.  Local versions of problems, though, are not found.

Explicitly, if a set def contains a reference like `source_file = setChain_rule/mo_ur_dr_5_20.pg` the problem won't be found because `setChain_rule/mo_ur_dr_5_20.pg` no longer exists in the course `templates` directory.  The fix is simple, prefix the problem path with the link to the model course.  In this case, `source_file = setChain_rule/mo_ur_dr_5_20.pg` becomes `source_file = Template_201-NYA/setChain_rule/mo_ur_dr_5_20.pg`.  In the case of 201-016, links like `source_file = set01_Background/ExponentialNotation.pg` will become `source_file = Template_201-016/set01_Background/ExponentialNotation.pg`.  I think I did a lot of these a year ago using some kind of search/replace.

My suggestion:

- [ ]  Merge the pull request
- [ ]  Clean up the local problem links in the set defs as above
    * I can do this or Brendan can do this
- [ ] Clean up the `templates` directory in 201-016 and add the link from `templates` into `model201-016` on the server
    * I can do this
* _Stretch goal_  Look at adding the local versions of problems into the CPL.

**Review by maharper — APPROVED** [review]
Looks great Brendan.

**Comment by qubd** [comment]
I just made the change in the local problem links.

For the future (N11 is next), do you mind if I put all the modified problems in one directory called "edits" or "modified"? That's how I did it in the course, and it would make this kind of find and replace work with the set definitions files much faster. I only did the work of separating the local problems into one directory per assignment because after glancing at another set of templates in here, I noticed that's how it was done, and figured maybe this was a requirement, or at least preferred.

**Comment by maharper** [comment]
> For the future (N11 is next), do you mind if I put all the modified problems in one directory called "edits" or "modified"? That's how I did it in the course, and it would make this kind of find and replace work with the set definitions files much faster. I only did the work of separating the local problems into one directory per assignment because after glancing at another set of templates in here, I noticed that's how it was done, and figured maybe this was a requirement, or at least preferred.

I think that you can organize the problem source files (`.pg`) however works best for you, you are doing the work after all.  There may be, however, some caveats.

**History**, or where did this organization structure come from?

Why is there structure like:
```
templates/
│   setAntiderivatives.def
│   setAsymptotes_and_continuity.def
├───setChain_rule/
│       mo_ur_dr_5_20.pg
│   setChain_rule.def
...
```
where of course `setChain_rule/mo_ur_dr_5_20.pg` is referenced in the set definitions file `setChain_rule.def`

The WW default would be the structure
```
templates/
├───local/
│       mo_ur_dr_5_20.pg
│   setAntiderivatives.def
│   setAsymptotes_and_continuity.def
│   setChain_rule.def
...
```
with a reference to `local/mo_ur_dr_5_20.pg` in `setChain_rule.def`

What we found is that after a while tweaking, editing the template assignments, correcting bugs locally etc, the structure started to look like
```
templates/
├───local/
│       mo_ur_dr_5_20.pg
│       local-edit-2.pg
│       local-edit-3.pg
│       random-name.pg
│   setAntiderivatives.def
│   setAsymptotes_and_continuity.def
│   setChain_rule.def
...
```
and we had no idea or record of where the problems in `local/` were used.  If particular, if I wanted to share the _Antiderivatives_ set from my NYA, I didn't know which problems from `local/` to send along with the set def.

We decided to store the local code for problems beside the referencing set definition file.  That way to share the _Chain rule_ set, I can highlight `setChain_rule/` and `setChain_rule.def` in the file manager, create archive and send off the resulting `.tgz` file.

This didn't solve all the problems, but did solve some of the problems.

**Current structure:** These days, much of our locally-authored code lives in a hierarchical directory structure in the CPL, tracked in git.  And the model course template assignments live in `Course-Templates` _together with the source code_ for any locally-authored or local versions of problems whose code is not in the CPL.  The above reasons for the `setChain_rule/` structure don't serve much purpose in `Course-Templates`.

There may be things to think about for organizing the `.pg` code in `model201-<nnn>`, though.  Will any code be moving to the CPL?  Perhaps it should be organized into a hierarchical CPL-like structure.  Is the local version a _bug-fix_ of an OPL problem?  Then we may want to track this problem, so we can revert to the OPL version of the problem once it is fixed in the OPL.  Is the local version a correction to notation or terminology?  These we would not expect to be corrected in the OPL, but we want to use our notation and terminology long-term.  Maybe we want to track updates or bug fixes to the original OPL problem.

As you can tell from the length of this comment, I've been thinking about this as I work on the 201-901 templates, but I'm not clear yet on best practice.  It would help me if I could discuss this with you in Zoom in the next week or so - maybe with @mtitcombe as well if she likes.

**TLDR:**  Where the `.pg` files go is up to you, but let's talk.

**Comment by mtitcombe** [comment]
https://github.com/cslmath/Champlain-Saint-Lambert-webwork-repo/pull/39#issuecomment-1021553025

Brendan, this is great! As per Malcolm's stretch goal of adding local problems to the CPL, perhaps you could think of the directory/subdirectory structure of where any local 016 problems would go.
